android/app/src/main/java/com/expensify/chat/CustomNotificationProvider.java

```diff
@@ -0,0 +1,234 @@
+
+    public CustomNotificationProvider(@NonNull Context context, @NonNull AirshipConfigOptions configOptions) {
+        super(context, configOptions);
+    }
+
+    @NonNull
+    @Override
+    protected NotificationCompat.Builder onExtendBuilder(@NonNull Context context, @NonNull NotificationCompat.Builder builder, @NonNull NotificationArguments arguments) {
+        super.onExtendBuilder(context, builder, arguments);
+        PushMessage message = arguments.getMessage();
+
+        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.M) {
+            return builder;
+        }
+
+        if (message.containsKey(PAYLOAD_KEY)) {
+            try {
+                JsonMap payload = JsonValue.parseString(message.getExtra(PAYLOAD_KEY)).optMap();
+
+                // Apply message style only for report comments
+                if (REPORT_COMMENT_TYPE.equals(payload.get(TYPE_KEY).getString())) {
+                        applyMessageStyle(builder, payload, arguments.getNotificationId());
+                }
+            } catch (Exception e) {
+                Log.e(TAG, "Failed to parse conversation", e);
+            }
+        }
+
+        return builder;
+    }
+
+    /**
+     * Applies the message style to the notification builder. It also takes advantage of the
+     * notification cache to build conversations.
+     *
+     * @param builder Notification builder that will receive the message style
+     * @param payload Notification payload, which contains all the data we need to build the notifications.
+     * @param notificationID Current notification ID
+     */
+    private void applyMessageStyle(NotificationCompat.Builder builder, JsonMap payload, int notificationID) {
+        int reportID = payload.get("reportID").getInt(-1);
+        NotificationCache notificationCache = findOrCreateNotificationCache(reportID);
+        JsonMap reportAction = payload.get("reportAction").getMap();
+        String name = reportAction.get("person").getList().get(0).getMap().get("text").getString();
+        String avatar = reportAction.get("avatar").getString();
+        String accountID = Integer.toString(reportAction.get("actorAccountID").getInt(-1));
+        String message = reportAction.get("message").getList().get(0).getMap().get("text").getString();
+        long time = reportAction.get("timestamp").getLong(0);
+        String roomName = payload.get("roomName") == null ? "" : payload.get("roomName").getString("");
+        String conversationTitle = "Chat with " + name;
+        if (!roomName.isEmpty()) {
+            conversationTitle = "#" + roomName;
+        }
+
+        IconCompat iconCompat = fetchIcon(avatar, FALLBACK_ICON_ID);
+        Person person = new Person.Builder()
+                .setIcon(iconCompat)
+                .setKey(accountID)
+                .setName(name)
+                .build();
+
+        if (!notificationCache.people.containsKey(accountID)) {
+            notificationCache.people.put(accountID, person);
+        }
```

Comment:
It looks like we're building a new person Object each time, rather than retrieving from cache? Or am I missing something?

Reply:
True! It should be fixed now.